src/bgw/scheduler.c

```diff
@@ -0,0 +1,386 @@
+/*
+ * This is a scheduler that takes background jobs and schedules them appropriately
+ *
+ * Limitations: For now the jobs are only loaded when the scheduler starts and are not
+ * updated if the jobs table changes
+ *
+ * TODO: right now jobs are not prioritized when slots available. Should prioritize in
+ * asc next_start order.
+ *
+*/
+#include <postgres.h>
+
+#include <miscadmin.h>
+#include <postmaster/bgworker.h>
+#include <storage/ipc.h>
+#include <storage/latch.h>
+#include <storage/lwlock.h>
+#include <storage/proc.h>
+#include <storage/shmem.h>
+#include <utils/jsonb.h>
+#include <utils/timestamp.h>
+#include <utils/snapmgr.h>
+#include <utils/memutils.h>
+#include <access/xact.h>
+#include <pgstat.h>
+
+#include "extension.h"
+#include "scheduler.h"
+#include "job.h"
+#include "job_stat.h"
+#include "version.h"
+#include "compat.h"
+#include "timer.h"
+
+#define SCHEDULER_APPNAME "TimescaleDB Background Worker Scheduler"
+
+#define LEAST_TIMESTAMP(left, right) \
+	(left < right ? left : right)
+
+TS_FUNCTION_INFO_V1(ts_bgw_scheduler_main);
+
+
+/* has to be global to shutdown jobs on exit */
+static List *scheduled_jobs = NIL;
+
+typedef enum JobState
+{
+	/* terminal state */
+	JOB_STATE_DISABLED,
+
+	/* next states: JOB_STATE_STARTED */
+	JOB_STATE_WAITING,
```

Comment:
Call this NOT_RUNNING or something since WAITING sounds like the job is running and waiting on something.

I think we decided on:

DISABLED
SCHEDULED
STARTED
TERMINATING -> SCHEDULED | DISABLED

Add some comments on what each state means, i.e., what the scheduler vs. job believes.

Reply:
Fixed